File: Examples/GenericStack/GenericStack.cs
// SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
// SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0

using System;
using System.Collections.Generic;

namespace Examples.GenericStack;

public sealed class Stack<T>
{
    private readonly List<T> items = [];

    public int Count => items.Count;
    public void Push(T value) => items.Add(value);
    public T Peek() => items[^1];

    public T Pop()
    {
        var value = items[^1];
        items.RemoveAt(items.Count - 1);
        return value;
    }
}

public static class GenericStack
{
    public static void Main()
    {
        var names = new Stack<string>();
        names.Push("Lexer");
        names.Push("Parser");
        names.Push("Core");
        Console.WriteLine(names.Peek());
        while (names.Count > 0)
        {
            Console.WriteLine(names.Pop());
        }
    }
}

File: Examples/GenericStack/GenericStack.java
// SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
// SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0

import java.util.ArrayList;
import java.util.List;

public final class GenericStack {
    private GenericStack() {}

    private static final class Stack<T> {
        private final List<T> items = new ArrayList<>();

        void push(T value) { items.add(value); }
        T peek() { return items.getLast(); }
        T pop() { return items.removeLast(); }
        int count() { return items.size(); }
    }

    public static void main(String[] args) {
        var names = new Stack<String>();
        names.push("Lexer");
        names.push("Parser");
        names.push("Core");
        System.out.println(names.peek());
        while (names.count() > 0) {
            System.out.println(names.pop());
        }
    }
}
